src/stim/util_top/reference_sample_tree.inl

```diff
@@ -0,0 +1,161 @@
+            child.repetitions = 1;
+            child.prefix_bits.insert(
+                child.prefix_bits.end(),
+                sim.measurement_record.storage.begin() + start_size,
+                sim.measurement_record.storage.begin() + end_size);
+        }
+        start_size = end_size;
+    };
+
+    for (size_t k = 0; k < reps; k++) {
+        for (const auto &inst : loop.operations) {
+            if (inst.gate_type == GateType::REPEAT) {
+                uint64_t repeats = inst.repeat_block_rep_count();
+                const auto& block = inst.repeat_block_body(loop);
+                flush_recorded_into_result();
+                result.suffix_children.push_back(do_loop_with_tortoise_hare_folding(block, repeats));
+                start_size = sim.measurement_record.storage.size();
+            } else {
+                sim.do_gate(inst);
+            }
+        }
+    }
+
+    flush_recorded_into_result();
+    return result;
+}
+
+static uint64_t max_feedback_lookback_in_loop(const Circuit &loop) {
+    uint64_t furthest_lookback = 0;
+    for (const auto &inst : loop.operations) {
+        if (inst.gate_type == GateType::REPEAT) {
+            furthest_lookback = std::max(furthest_lookback, max_feedback_lookback_in_loop(inst.repeat_block_body(loop)));
+        } else {
+            auto f = GATE_DATA[inst.gate_type].flags;
+            if ((f & GateFlags::GATE_CAN_TARGET_BITS) && (f & GateFlags::GATE_TARGETS_PAIRS)) {
+                // Feedback-capable operation. Check for any measurement record targets.
+                for (auto t : inst.targets) {
+                    if (t.is_measurement_record_target()) {
+                        furthest_lookback = std::max(furthest_lookback, (uint64_t)-t.rec_offset());
+                    }
+                }
+            }
+        }
+    }
+    return furthest_lookback;
+}
+
+template <size_t W>
+ReferenceSampleTree CompressedReferenceSampleHelper<W>::do_loop_with_tortoise_hare_folding(const Circuit &loop, uint64_t reps) {
+    if (reps < 10) {
+        return do_loop_with_no_folding(loop, reps);
+    }
+
+    ReferenceSampleTree result;
+    result.repetitions = 1;
+
+    CompressedReferenceSampleHelper<W> tortoise(sim);
+    CompressedReferenceSampleHelper<W> hare(std::move(sim));
+    uint64_t max_feedback_lookback = max_feedback_lookback_in_loop(loop);
+    uint64_t tortoise_steps = 0;
+    uint64_t hare_steps = 0;
+    while (hare_steps < reps) {
+        hare_steps++;
+        result.suffix_children.push_back(hare.do_loop_with_no_folding(loop, 1));
+        assert(result.suffix_children.size() == hare_steps);
+
+        if (tortoise.in_same_recent_state_as(hare, max_feedback_lookback, hare_steps < 10)) {
+            break;
+        }
+
+        // Tortoise advances half as quickly.
+        if (hare_steps & 1) {
+            tortoise_steps++;
+            tortoise.do_loop_with_no_folding(loop, 1);
+        }
+    }
+
+    if (hare_steps == reps) {
```

Comment:
When that happens it's acceptable to treat it not as a loop, because it would output a `REPEAT 1 { ... }` that then immediately gets flattened by the simplification code into the same thing as this produces.

Reply:
I was thinking that it was essentially equivalent (no need for the follow up extension below). If that's always the case, that's fine. But, you should update the comment on the next line to cover that.